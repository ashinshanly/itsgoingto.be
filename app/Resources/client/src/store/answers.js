import { createSelector } from 'reselect'
import { prop, adjust, nth, compose, not, equals, length, remove, omit, is, merge, find,
         isEmpty, slice, findLastIndex, when, subtract, __, gt, trim, map, isNil, propEq } from 'ramda'

// TODO : Update answers to reference by identifier and index

// ------------------------------------
// Constants
// ------------------------------------
export const ANSWER_ADD           = 'ANSWER_ADD'
export const ANSWER_UPDATE        = 'ANSWER_UPDATE'
export const ANSWERS_UPDATE       = 'ANSWERS_UPDATE'
export const ANSWER_REMOVE        = 'ANSWER_REMOVE'
export const ANSWERS_REMOVE_AFTER = 'ANSWERS_REMOVE_AFTER'
export const ANSWERS_CLEAR        = 'ANSWERS_CLEAR'

// ------------------------------------
// Selectors
// ------------------------------------
/**
 * Select answers from the state
 *
 * @param  {object} state      App state
 * @param  {string} identifier Poll identifier
 *
 * @return {array}             Answers
 */
export const answersSelector = (state, identifier = '') => prop('answers')(state)

/**
 * Returns the max answer index using the answersSelector
 *
 * @param  {object} state      App state
 * @param  {string} identifier Poll identifier
 *
 * @return {integer}           Last index
 */
export const maxAnswerSelector = createSelector(
  answersSelector,
  answers => findLastIndex(compose(not, isEmpty, trim))(answers)
)

/**
 * Get the answer at a given index
 *
 * @param  {object}  state  App state
 * @param  {integer} index  Index of answer
 *
 * @return {object}         Answer object
 */
export const answerSelector = (state, index) => compose(nth(index), answersSelector)(state)

/**
 * Get if the poll has an answer at index
 *
 * @param  {object}  state  App state
 * @param  {integer} index  Index of answer
 *
 * @return {bool}
 */
export const hasAnswerSelector = (state, index) => compose(not, equals(0), length, trim, answerSelector)(state, index)

/**
 * Can the poll be submitted, return true if the poll has more than 2 answers
 *
 * @param  {object} state  App state
 *
 * @return {bool}
 */
export const canSubmitPollSelector = (state) => compose(gt(__, 2), length, answersSelector)(state)

// ------------------------------------
// Actions
// ------------------------------------
/**
 * Add an answer to the state
 *
 * @return {object} dispatchable object
 */
export const addAnswer = () => ({
  type : ANSWER_ADD
})

/**
 * Update the answer in the state the dispatch to add or remove answers
 *
 * @param  {integer} index Answer index
 * @param  {String}  value Answer text
 *
 * @return {Function}      redux-thunk callable function
 */
export const updateAnswer = (index, value = '') => (dispatch, getState) => {
  let hadAnswer = hasAnswerSelector(getState(), index)
  dispatch({
    type  : ANSWER_UPDATE,
    index : index,
    text  : value
  })
  let hasAnswer = hasAnswerSelector(getState(), index)
  let countAnswers = length(answersSelector(getState()))

  if (index === (countAnswers - 1) && hasAnswer && !hadAnswer) {
    dispatch(addAnswer())
  } else if (hadAnswer && !hasAnswer) {
    dispatch(removeAfterAnswer(maxAnswerSelector(getState()) + 1))
  }

  return Promise.resolve()
}

/**
 * Update all the answers in the state
 *
 * @param  {array}  answers Array of answers
 *
 * @return {object}         dispatchable object
 */
export const updateAnswers = (answers = []) => ({
  type    : ANSWERS_UPDATE,
  answers : when(
    compose(not, isNil),
    map(
      when(
        is(Object),
        omit(['poll'])
      ),
    )
  )(answers)
})

/**
 * Remove an answer at the index
 *
 * @param  {integer} index Index of the answer
 *
 * @return {object}         dispatchable object
 */
export const removeAnswer = (index) => ({
  type  : ANSWER_REMOVE,
  index
})

/**
 * Remove answers after the index
 *
 * @param  {integer} index Index after which to remove answers
 *
 * @return {object}         dispatchable object
 */
export const removeAfterAnswer = (index) => ({
  type  : ANSWERS_REMOVE_AFTER,
  index
})

/**
 * Clear all the answers from the state
 *
 * @return {object}         dispatchable object
 */
export const clearAnswers = () => ({
  type : ANSWERS_CLEAR
})

export const actions = {
  addAnswer,
  updateAnswer,
  updateAnswers,
  removeAnswer,
  removeAfterAnswer,
  clearAnswers
}

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  // Add a new answer to the state
  [ANSWER_ADD]           : (previousState, action) => [...previousState, ''],
  // Update an answer in the state
  [ANSWER_UPDATE]        : (previousState, action) => adjust(() => action.text, action.index, previousState),
  // Update all the answers in the state
  [ANSWERS_UPDATE]       : (previousState, action) => map(
      answer => when(
        is(Object),
        merge(find(propEq('id', answer.id), previousState))
      )(answer)
    )(action.answers),
  // Remove an answer in the state
  [ANSWER_REMOVE]        : (previousState, action) =>
    when(
      compose(not, equals(action.index), subtract(__, 1), length),
      remove(action.index, 1)
    )(previousState),
  // Remove the answers after the index in the state
  [ANSWERS_REMOVE_AFTER] : (previousState, action) => slice(0, action.index + 1, previousState),
  // Clear all the answers from the state
  [ANSWERS_CLEAR]        : (previousState, action) => []
}

// ------------------------------------
// Reducer
// ------------------------------------
/**
 * Initial state for this store component
 */
const initialState = []

/**
 * The reducer for this store component
 *
 * @param  {State} state   The current state
 * @param  {object} action The action to perform on the state
 *
 * @return {State}         The modified state
 */
export default function answersReducer (state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]
  return handler ? handler(state, action) : state
}
