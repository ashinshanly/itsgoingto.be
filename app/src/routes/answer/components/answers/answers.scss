@import "~bourbon/core/bourbon";
@import "variables";

.answer-container {
  text-align: center;
  margin-bottom: 10em;

  @media(max-width: $screen-sm-max) {
    margin-top: 8em;
  }

  .options {
    display: inline-block;
    text-align: left;
  }
}
