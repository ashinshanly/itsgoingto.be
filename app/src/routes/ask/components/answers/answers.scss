.input-label-answers {
  margin-bottom: 0.5em;
}
