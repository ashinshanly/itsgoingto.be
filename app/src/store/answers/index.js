import {
  adjust, compose, not, equals, length, remove, is, merge, find,
  slice, when, subtract, __, map, propEq
} from 'ramda'
import {
  ANSWER_ADD, ANSWER_UPDATE, ANSWERS_UPDATE, ANSWER_REMOVE, ANSWERS_REMOVE_AFTER, ANSWERS_CLEAR
} from 'store/answers/actions'

// TODO : Update answers to reference by identifier and index

// ------------------------------------
// Action Handlers
// ------------------------------------
const ACTION_HANDLERS = {
  // Add a new answer to the state
  [ANSWER_ADD]           : (previousState) => [...previousState, ''],
  // Update an answer in the state
  [ANSWER_UPDATE]        : (previousState, action) => adjust(() => action.text, action.index, previousState),
  // Update all the answers in the state
  [ANSWERS_UPDATE]       : (previousState, action) =>
    map(
      (answer) => when(
        is(Object),
        merge(find(propEq('id', answer.id), previousState))
      )(answer)
    )(action.answers),
  // Remove an answer in the state
  [ANSWER_REMOVE]        : (previousState, action) =>
    when(
      compose(not, equals(action.index), subtract(__, 1), length),
      remove(action.index, 1)
    )(previousState),
  // Remove the answers after the index in the state
  [ANSWERS_REMOVE_AFTER] : (previousState, action) => slice(0, action.index + 1, previousState),
  // Clear all the answers from the state
  [ANSWERS_CLEAR]        : () => []
}

// ------------------------------------
// Reducer
// ------------------------------------
/**
 * Initial state for this store component
 */
const initialState = []

/**
 * The reducer for this store component
 *
 * @param  {State} state   The current state
 * @param  {object} action The action to perform on the state
 *
 * @return {State}         The modified state
 */
export default function answersReducer (state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]

  return handler ? handler(state, action) : state

  // switch (action.type) {
  //   case POLL_UPDATE:
  //    return 
  //   default:
  //     return state
  // }
}
