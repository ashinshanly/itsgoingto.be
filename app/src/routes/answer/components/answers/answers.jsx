import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import PropTypes from 'prop-types'
import { prop, compose, contains, when, isNil } from 'ramda'
import { fetchResponses, postResponse, fetchPoll, APIError } from 'store/api'
import Spinner from 'components/spinner'
import Answer from '../answer'
import './answers.scss'

class Answers extends React.Component {
  updateAnswers = () => {
    const { poll, updateResponses, fetchPoll } = this.props

    if (poll.ended || poll.deleted) {
      clearInterval(this.answersUpdater)
    } else {
      updateResponses()
        .then((response) => {
          if (response instanceof APIError) {
            fetchPoll()
          }
        })
    }
  }

  componentDidMount = () => {
    this.answersUpdater = setInterval(
      this.updateAnswers,
      5000
    )
  }

  componentWillUnmount = () => {
    clearInterval(this.answersUpdater)
  }

  answerChecked = (answer) =>
    compose(contains(answer.id), when(isNil, () => []), prop('userResponses'))(this.props.poll)

  render () {
    const { poll, answers, userResponded, postResponse, totalResponses, viewOnly } = this.props

    return (
      <div className='container answer-container'>
        {answers.length === 0 &&
          <div className={'spinner-container center-text'}>
            <Spinner />
          </div>
        }
        <div className={'options' + (userResponded || viewOnly || poll.ended ? ' show-results' : '')}>
          {answers.map((answer, index) =>
            <Answer
              key={index}
              index={index}
              type={poll.multipleChoice ? 'checkbox' : 'radio'}
              answer={answer}
              poll={poll}
              viewOnly={viewOnly}
              checked={this.answerChecked(answer)}
              postResponse={postResponse}
              totalResponses={totalResponses} />
          )}
        </div>
      </div>
    )
  }
}

Answers.propTypes = {
  answers         : PropTypes.array.isRequired,
  poll            : PropTypes.object.isRequired,
  totalResponses  : PropTypes.number,
  userResponded   : PropTypes.bool.isRequired,
  viewOnly        : PropTypes.bool.isRequired,
  updateResponses : PropTypes.func.isRequired,
  postResponse    : PropTypes.func.isRequired,
  fetchPoll       : PropTypes.func.isRequired
}

Answers.defaultProps = {
  totalResponses : 0
}

const mapDispatchToProps = (dispatch, props) => ({
  updateResponses : () => dispatch(fetchResponses(props.params.identifier)),
  postResponse    : (id) => dispatch(postResponse(id, props.params.identifier)),
  fetchPoll       : () => dispatch(fetchPoll(props.params.identifier))
})

export default withRouter(connect(null, mapDispatchToProps)(Answers))
