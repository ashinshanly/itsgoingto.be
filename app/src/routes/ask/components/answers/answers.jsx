import React from 'react'
import PropTypes from 'prop-types'
import { length } from 'ramda'
import Answer from './answer'
import './answers.scss'

export const Answers = ({ hasQuestion, answers }) => (
  <div className={'answers hideable' + (hasQuestion ? '' : ' gone')}>
    <label className='input-label input-label-answers'>Add some responses</label>
    {answers.map((answer, index) =>
      <Answer key={index} index={index} text={answer.answer ? answer.answer : answer} />
    )}
    <Answer key={length(answers)} index={length(answers)} disabled />
  </div>
)

Answers.propTypes = {
  hasQuestion : PropTypes.bool.isRequired,
  answers     : PropTypes.array.isRequired
}

export default Answers
