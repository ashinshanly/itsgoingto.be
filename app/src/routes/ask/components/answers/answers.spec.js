/* global expect, jest */
import React from 'react'
import { shallow } from 'enzyme'
import { Answers } from './answers'

const props = {
  hasQuestion : true,
  answers     : [],
}

describe('(Route) Ask', () => {
  describe('(Component) Answers', () => {
    describe('(Render)', () => {
      describe('when hasQuestion is true', () => {
        it('matches snapshot', () => {
          const wrapper = shallow(<Answers {...props} hasQuestion={true} />)
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('when hasQuestion is false', () => {
        it('matches snapshot', () => {
          const wrapper = shallow(<Answers {...props} hasQuestion={false} />)
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('with answers as strings', () => {
        const answers = [
          'Answer 1',
          'Answer 2',
          'Answer 3'
        ]
        it('matches snapshot', () => {
          const wrapper = shallow(<Answers {...props} answers={answers} />)
          expect(wrapper).toMatchSnapshot()
        })
      })

      describe('with answers as objects', () => {
        const answers = [
          { answer: 'Answer 4' },
          { answer: 'Answer 5' },
          { answer: 'Answer 6' }
        ]
        it('matches snapshot', () => {
          const wrapper = shallow(<Answers {...props} answers={answers} />)
          expect(wrapper).toMatchSnapshot()
        })
      })
    })
  })
})
