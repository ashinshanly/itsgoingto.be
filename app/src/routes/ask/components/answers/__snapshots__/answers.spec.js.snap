// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Ask (Component) Answers (Render) when hasQuestion is false matches snapshot 1`] = `
<div
  className="answers hideable gone"
>
  <label
    className="input-label input-label-answers"
  >
    Add some responses
  </label>
  <Connect(Answer)
    disabled={true}
    index={0}
    key="0"
  />
</div>
`;

exports[`(Route) Ask (Component) Answers (Render) when hasQuestion is true matches snapshot 1`] = `
<div
  className="answers hideable"
>
  <label
    className="input-label input-label-answers"
  >
    Add some responses
  </label>
  <Connect(Answer)
    disabled={true}
    index={0}
    key="0"
  />
</div>
`;

exports[`(Route) Ask (Component) Answers (Render) with answers as objects matches snapshot 1`] = `
<div
  className="answers hideable"
>
  <label
    className="input-label input-label-answers"
  >
    Add some responses
  </label>
  <Connect(Answer)
    index={0}
    key="0"
    text="Answer 4"
  />
  <Connect(Answer)
    index={1}
    key="1"
    text="Answer 5"
  />
  <Connect(Answer)
    index={2}
    key="2"
    text="Answer 6"
  />
  <Connect(Answer)
    disabled={true}
    index={3}
    key="3"
  />
</div>
`;

exports[`(Route) Ask (Component) Answers (Render) with answers as strings matches snapshot 1`] = `
<div
  className="answers hideable"
>
  <label
    className="input-label input-label-answers"
  >
    Add some responses
  </label>
  <Connect(Answer)
    index={0}
    key="0"
    text="Answer 1"
  />
  <Connect(Answer)
    index={1}
    key="1"
    text="Answer 2"
  />
  <Connect(Answer)
    index={2}
    key="2"
    text="Answer 3"
  />
  <Connect(Answer)
    disabled={true}
    index={3}
    key="3"
  />
</div>
`;
